harvestor/config/config.go

```diff
@@ -0,0 +1,119 @@
+// harvestor is an application for recursively traverse directories and find media files.
+// send each file of our interest to to object-store-api and map it back to the metadata
+// For more information please refer to https://github.com/AAFC-BICoE/object-store-harvestor/blob/dev/doc/design.md
+
+// TODO For now we do not modify config while the app is running
+// viper has OnConfigChange event which we can trigger to reload our config during the app run
+// Do not see yet this case. May be later on with UI for the user
+
+package config
+
+import (
+	"fmt"
+	"github.com/spf13/viper"
+	"log"
+	"path/filepath"
+	"strings"
+)
+
+type Configuration struct {
+	Database   DatabaseConfiguration   // SQLite DB config
+	Walker     FileWalkerConfiguration // Media File walker config
+	HttpClient HttpClientConfiguration // Http Client config
+	Logger     LoggerConfiguration     // Logger config
+	App        AppConfiguration        // App config
+}
+
+// Validation helpers
+var (
+	ValidConfigFileExtensions                    = []string{"yml"}
+	validConfigFileExtensionsMap map[string]bool = make(map[string]bool)
+)
+
+// define global empty configuration
+var conf Configuration
+
+func Load(filename string) {
+	err := readFromFile(filename)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func GetConf() Configuration {
+	return conf
+}
+
+func GetLoggerLevel() string {
+	return strings.ToLower(conf.Logger.Level)
+}
+
+func readFromFile(filename string) error {
+	// get path and file
+	path, file := filepath.Split(filename)
+
+	// validation on supported extention
+	if !isValidConfigFile(file) {
+		err := fmt.Errorf("config file:(%q) is not supported with extention:(%q)", file, getFileExtension(file))
+		return err
+	}
+
+	// define file name
+	name := getFileName(file)
+	// define file extension
+	extension := getFileExtension(file)
+	// Debug for now
+	log.Println("---------------------- C o n f i g ----------------------------")
+	log.Println("||| config file path :", path)
+	log.Println("||| config file name :", name)
+	log.Println("||| config file extension :", extension)
+	log.Println("---------------------------------------------------------------")
+
+	// init new viper
+	v := viper.New()
+	// passing file to viper
+	// config file name without extension
+	v.SetConfigName(name)
+	// config file extension
+	v.SetConfigType(extension)
+	// config file path
+	v.AddConfigPath(path)
+	// in case we have .env
+	// read value ENV variable
+	v.AutomaticEnv()
+
+	// Reading from yml file
+	err := v.ReadInConfig()
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal to predefined struct
+	err = v.Unmarshal(&conf)
+	if err != nil {
+		return err
+	}
+
+	// returning Configuration
+	return nil
+}
+
+// check if the config type supported
+func isValidConfigFile(file string) bool {
+	for _, ext := range ValidConfigFileExtensions {
```

Comment:
I'm not a "slice" expert but you can possibly change `ValidConfigFileExtensions` for a map and use `value, exist := ValidConfigFileExtensions[ext]` and return `exist` ?

Reply:
I was thinking to support multiple formats in the future, such as yaml or yml. The point was to avoid json or xml for example or limited to only to what we can accept. viper library can read pretty much any of the existing formats.
Sure, I can check only against one value. I will change the code

Reply:
I changed the code to simplify validation on config file extension. Please take a look. Thank you